Stratis.Bitcoin/BlockPulling/LookaheadBlockPuller.cs

```diff
@@ -150,7 +150,11 @@ public int MaxBufferedSize
         /// <summary>Current number of bytes that unconsumed blocks are occupying.</summary>
         private long currentSize;
 
+        /// <summary>Lock object to protect access to <see cref="downloadedCounts"/>.</summary>
+        private object downloadedCountsLock = new object();
+
         /// <summary>Maintains the statistics of number of downloaded blocks. This is used for calculating new actualLookahead value.</summary>
+        /// <remarks>All access to this object has to be protected by <see cref="downloadedCountsLock"/>.</remarks>
```

Comment:
Mostly because of this:

```
        public decimal MedianDownloadCount
        {
            get
            {
                lock (this.downloadedCountsLock)
                {
                    if (this.downloadedCounts.Count == 0)
                        return decimal.One;
                    return (decimal)GetMedian(this.downloadedCounts);
                }
            }
        }
```

If I had a ConcurrentBag there, the race would not be prevented. I could still get .Count != 0 in the first IF line, then having the collection changed by another thread and then calling GetMedian with empty collection and get that exception again. Unless I changed the logic of course, but why would I do that. 

In general, using Concurrent collections is not saving your from anything except for trivial single operations. Once you have conditional operation like this, Concurrent stuff is not helping. This is why I try to avoid Concurrent collections completely as they tend to misguide you - giving you a false sense of security.